support/src/main/java/org/jenkinsci/plugins/workflow/support/steps/MilestoneStepExecution.java

```diff
@@ -0,0 +1,443 @@
+            Iterator<Integer> it2 = inSight.iterator();
+            while (it2.hasNext()) {
+                Integer number = it2.next();
+                if (job.getBuildByNumber(number) == null) {
+                    // Deleted at some point but did not properly clean up from exit(…).
+                    LOGGER.log(WARNING, "Cleaning up apparently deleted {0}#{1}", new Object[] {jobName, number});
+                    it2.remove();
+                }
+            }
+        }
+        // TODO: remove milestones that no longer exist (because the step or the job was deleted)
+    }
+
+    @SuppressWarnings("unchecked")
+    private static synchronized void load() {
+        if (milestonesByOrdinalByJob == null) {
+            milestonesByOrdinalByJob = new TreeMap<String, Map<Integer, Milestone>>();
+            try {
+                XmlFile configFile = getConfigFile();
+                if (configFile.exists()) {
+                    milestonesByOrdinalByJob = (Map<String, Map<Integer, Milestone>>) configFile.read();
+                }
+            } catch (IOException x) {
+                LOGGER.log(WARNING, null, x);
+            }
+            LOGGER.log(Level.FINE, "load: {0}", milestonesByOrdinalByJob);
+        }
+    }
+
+    private static synchronized void save() {
+        try {
+            getConfigFile().write(milestonesByOrdinalByJob);
+        } catch (IOException x) {
+            LOGGER.log(WARNING, null, x);
+        }
+        LOGGER.log(Level.FINE, "save: {0}", milestonesByOrdinalByJob);
+    }
+
+    private static XmlFile getConfigFile() throws IOException {
+        Jenkins j = Jenkins.getInstance();
+        if (j == null) {
+            throw new IOException("Jenkins is not running"); // do not use Jenkins.getActiveInstance() as that is an ISE
+        }
+        return new XmlFile(new File(j.getRootDir(), MilestoneStep.class.getName() + ".xml"));
+    }
+
+    @Extension
+    public static final class Listener extends RunListener<Run<?,?>> {
+        @Override public void onCompleted(Run<?,?> r, TaskListener listener) {
+            if (!(r instanceof FlowExecutionOwner.Executable) || ((FlowExecutionOwner.Executable) r).asFlowExecutionOwner() == null) {
+                return;
+            }
+            exit(r);
+        }
+    }
+
+    private static final class Milestone {
+
+        /**
+         * Milestone ordinal.
+         */
+        final Integer ordinal;
+
+        /**
+         * Numbers of builds that passed this milestone but haven't passed the next one.
+         */
+        final Set<Integer> inSight = new TreeSet<Integer>();
+
+        /**
+         * Last build that passed through the milestone, or null if none passed yet.
+         */
+        @CheckForNull
+        Integer lastBuild;
+
+        /**
+         * Last build extenalizable ID hat passed through the milestone, or null if none passed yet.
+         */
+        @CheckForNull
+        String lastBuildExternalizableId;
```

Comment:
:ant: redundant to store this when we have `lastBuild`; the ID simply consists of the job name (which we already know) and the build number (for Jenkins 1.597+, which we must be running for this new step).